File: rust_dep/src/sir.rs


pub struct Parameters {
    pub Beta: Vec<f32>,
    pub Gamma: Vec<f32>,
    pub Alpha: Vec<f32>,
}

pub struct SIR {
    pub S: f32,
    pub I: f32,
    pub R: f32,
}

pub struct Results {
    pub S: Vec<f32>,
    pub I: Vec<f32>,
    pub R: Vec<f32>,
}

impl Results {
    fn new() -> Self {
        return Results {
            S: Vec::new(),
            I: Vec::new(),
            R: Vec::new()
        }
    }
}

fn Sdot(state: &SIR, params: &Parameters , i: &usize) -> f32 {
    - params.Beta[*i] * (( (state.S * state.I) / (state.S + state.I + state.R) ) as f32) +
    params.Alpha[*i] * (state.R as f32)
}

fn Idot(state: &SIR, params: &Parameters , i: &usize) -> f32 {
    params.Beta[*i] * (( (state.S * state.I) / (state.S + state.I + state.R) ) as f32) -
    params.Gamma[*i] * ( state.I as f32)
}

fn Rdot(state: &SIR, params: &Parameters , i: &usize) -> f32 {
    params.Gamma[*i] * ( state.I as f32) - params.Alpha[*i] * (state.R as f32)
}

fn step(state: &SIR, params: &Parameters, i: &usize, dt: &f32) -> SIR {
    let S = state.S + Sdot(state, params, i) * dt;
    let I = state.I + Idot(state, params, i) * dt;
    let R = state.R + Rdot(state, params, i) * dt;
    return SIR {S, I, R}

}

fn capture_state(state: &SIR, results: &mut Results) {
    results.S.push(state.S);
    results.I.push(state.I);
    results.R.push(state.R);
}


pub fn run(runtime: f32, dt: f32, initial_state: SIR, params: Parameters) -> Results {
    let mut i = 0 as usize;
    let mut t = 0 as f32;
    let mut state = initial_state;
    let mut res = Results::new();
    while t < runtime - dt {
        state = step(&state, &params, &i, &dt);
        capture_state(&state, &mut res);
        i += 1;
        t += dt;
    }
    res
}

File: src/sir.py
import numpy as np
import pandas as pd


def get_lambda(contact_matrix, pop):
    M_diag_p = np.matmul( contact_matrix, (np.diag(pop)/pop.sum()).astype(float) )
    lambda_ = np.linalg.eig(M_diag_p)[0].max() # dominant eigenvalue
    return lambda_


class SIR_arr:
    def __init__(self, S0, I0, R0, Beta, Gamma, dt, group_labels=None):
        if not isinstance(S0, list):
            S0 = [S0]
        if not isinstance(I0, list):
            I0 = [I0]
        if not isinstance(R0, list):
            R0 = [R0]
        assert all([Sg >= 0 for Sg in S0]) and all([Rg >= 0 for Rg in R0]) and all([Ig >= 0 for Ig in I0]), 'Starting population must be greater than or equal to zero for all compartments.'
        
        if group_labels is not None:
            self.group_labels = group_labels
        else:
            self.group_labels = [str(i) for i in range(len(S0))]
            
        assert len(self.group_labels) == len(S0) and \
            len(self.group_labels) == len(I0) and \
            len(self.group_labels) == len(R0), 'Initial population values must align with (age) groups.'
        
        self.St = [S0]
        self.It = [I0]
        self.Rt = [R0]
        
        if not isinstance(Beta, list):
            self.Beta = [[Beta]*self.n_groups]*self.n_groups
        elif len(Beta) == len(Beta[0]) and len(Beta) == self.n_groups:
            self.Beta = Beta
        else:
            raise Exception("Beta must be a square matrix or constant.")
            
        if not isinstance(Gamma, list):
            self.Gamma = [Gamma]*self.n_groups
        elif len(Gamma) == self.n_groups:
            self.Gamma = Gamma
        else:
            raise Exception(f"Gamma must be a or constant or a list of size (n_groups == {self.n_groups}).")
        
        
        self.time = 0
        self.dt = dt
        self.N = sum(S0) + sum(I0) + sum(R0)
        
    @property
    def time_vector(self):
        return np.arange(0, self.time + self.dt, self.dt)[:len(self.S_timecourse())]
    
    @property
    def n_groups(self):
        return len(self.group_labels)
    
    def S_timecourse(self, group=None):
        if group is None:
            return [sum(S) for S in self.St]
        else:
            return [S[group] for S in self.St]
        
    def I_timecourse(self, group=None):
        if group is None:
            return [sum(I) for I in self.It]
        else:
            return [I[group] for I in self.It]
        
    def R_timecourse(self, group=None):
        if group is None:
            return [sum(R) for R in self.Rt]
        else:
            return [R[group] for R in self.Rt]

    def run(self, runtime):
        for _ in np.arange(self.time, self.time+runtime, self.dt):
            self.step()

    def step(self):
        S_new = [self.St[-1][i] + self.dt*self.Sdot(group=i) for i in range(self.n_groups)]
        I_new = [self.It[-1][i] + self.dt*self.Idot(group=i) for i in range(self.n_groups)]
        R_new = [self.Rt[-1][i] + self.dt*self.Rdot(group=i) for i in range(self.n_groups)]

        self.St.append(S_new)
        self.It.append(I_new)
        self.Rt.append(R_new)

        self.time += self.dt

    def Sdot(self, group=0):
        return - self.St[-1][group] * sum([self.Beta[group][j] * self.It[-1][j] for j in range(self.n_groups)]) / self.N

    def Idot(self, group=0):
        return self.St[-1][group] * sum([self.Beta[group][j] * self.It[-1][j] for j in range(self.n_groups)]) / self.N \
            - self.Gamma[group] * self.It[-1][group]

    def Rdot(self, group=0):
        return self.Gamma[group] * self.It[-1][group]